### PR TITLE
add remaining root_validators (depend on PR106)

### PR DESCRIPTION
**Pull request by feilin-flexcompute**
- [x] add get_raw_dict() function to exclude _type and comments from dict(Model)
    - [x] facilitate type checking using isinstance() in validations.py
- [x] move all remaining root_validators of Flow360Params to validations.py
- [x] add unit tests for all newly added validators (8 new files test_validator_*.py)
- [x] move validator-related tests from test_param_boundary.py to test_validator_boundary.py (cleaner tests)
- [x] typing of axis-type variable, e.g. sliceNormal, axisOfRotation, is changed to Axis

### CODE REVIEW

File: flow360/component/flow360_params/validations.py [hunk 1]
```diff
@@ -37,3 +49,176 @@ def _check_duplicate_boundary_name(values):
...
+            )
+        if "kOmegaSST_DDES" in output_fields and not (model_type == "kOmegaSST" and run_ddes):
+            raise ValueError(
+                "kOmegaSST_DDES output can only be specified with kOmegaSST turbulence model and DDES turned on"
+            )
+    return values
+
+
+def _validate_cht_has_heat_transfer_zone(values):
+    heat_equation_solver = values.get("heat_equation_solver")
```
Comment (maciej-flexcompute) [inline flow360/component/flow360_params/validations.py:97]: we need to make sure this validator runs after alias inference, i.e., after heatEquationSolver is converted to heat_equation_solver. Not sure at  which stage alias are applied.

File: flow360/component/flow360_params/validations.py
Comment (maciej-flexcompute) [inline flow360/component/flow360_params/validations.py:1]: we need tests for these validators

File: flow360/component/flow360_params/validations.py (hunk 1 above)
Comment (feilin-flexcompute) [reply flow360/component/flow360_params/validations.py:97]: I will add some unit tests to confirm

File: flow360/component/flow360_params/validations.py
Comment (feilin-flexcompute) [reply flow360/component/flow360_params/validations.py:1]: sure thing

File: flow360/component/flow360_params/params_base.py [hunk 2]
```diff
@@ -493,6 +493,16 @@ def _convert_dimensions_to_solver(
...
 
+    def get_raw_dict(self) -> dict:
+        """
+        get raw dictionary where sub-models are not converted to dictionaries,
+        without comments and _type.
+        """
+        self_copy = dict(self.copy(deep=True))
+        for key in [COMMENTS, TYPE_TAG_STR]:
+            self_copy.pop(key, None)
+        return dict(self_copy)
```
[the remark this answers is not in the document]
Comment (feilin-flexcompute) [reply flow360/component/flow360_params/params_base.py]: this function is removed.

File: flow360/component/flow360_params/validations.py [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
+from .time_stepping import UnsteadyTimeStepping
+from .volume_zones import HeatTransferVolumeZone
 
 
 def _check_tri_quad_boundaries(values):
     boundaries = values.get("boundaries")
     boundary_names = []
     if boundaries is not None:
-        boundary_names = list(boundaries.dict().keys())
+        boundary_names = list(boundaries.get_raw_dict().keys())
```
[the remark this answers is not in the document]
Comment (feilin-flexcompute) [reply flow360/component/flow360_params/validations.py]: fixed.

File: tests/params/test_validator_bet_disks.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+assertions = unittest.TestCase("__init__")
+
+
+@pytest.fixture(autouse=True)
+def change_test_dir(request, monkeypatch):
+    monkeypatch.chdir(request.fspath.dirname)
+
+
+def test_bet_disks():
```
Comment (maciej-flexcompute) [inline tests/params/test_validator_bet_disks.py:74]: lets add some tests also when using files, for example:
```python
param = Flow360Params('case_with_incorrect_bet.json')
```
and see if it throws the same error.

File: flow360/component/flow360_params/validations.py [hunk 5]
```diff
@@ -37,3 +53,375 @@ def _check_duplicate_boundary_name(values):
...
+
+    surface_output_fields = []
+    surface_output = values.get("surface_output")
+    if surface_output is not None:
+        surface_output_fields = surface_output.output_fields
+    if "wallFunctionMetric" in surface_output_fields and (not has_wall_function_boundary):
+        raise ValueError(
+            "'wallFunctionMetric' in 'surfaceOutput' is only valid for 'WallFunction' boundary type."
+        )
+    return values
```
[the remark this answers is not in the document]
Comment (awkrupka) [reply flow360/component/flow360_params/validations.py]: done on [a081f87](https://github.com/flexcompute/Flow360/pull/111/commits/a081f87a58d5301cda0ca943de45db464501aae9)

File: flow360/component/flow360_params/validations.py [hunk 6]
```diff
@@ -37,3 +53,375 @@ def _check_duplicate_boundary_name(values):
...
+            raise ValueError(f"Periodic pair for patch {boundary_key} is not specified.")
+
+    return values
+
+
+def _check_bet_disks_alphas_in_order(bet_disks) -> NoReturn:
+    for index, disk in enumerate(bet_disks):
+        alphas = disk.alphas
+        if alphas != sorted(alphas):
+            raise ValueError(f"BET Disk {index}: alphas are not in increasing order.")
```
[the remark this answers is not in the document]
Comment (feilin-flexcompute) [reply flow360/component/flow360_params/validations.py]: done

File: flow360/component/flow360_params/validations.py [hunk 7]
```diff
@@ -37,3 +53,375 @@ def _check_duplicate_boundary_name(values):
...
+        if has_duplicate:
+            raise ValueError(
+                f"BET Disk {index} has duplicated radius at {duplicated_radius} in chords."
+            )
+        twists = disk.twists
+        duplicated_radius, has_duplicate = _check_has_duplicate_in_one_radial_list(twists)
+        if has_duplicate:
+            raise ValueError(
+                f"BET Disk {index} has duplicated radius at {duplicated_radius} in twists."
+            )
```
[the remark this answers is not in the document]
Comment (feilin-flexcompute) [reply flow360/component/flow360_params/validations.py]: done

File: flow360/component/flow360_params/validations.py [hunk 8]
```diff
@@ -37,3 +53,375 @@ def _check_duplicate_boundary_name(values):
...
+
+
+def _check_bet_disks_number_of_defined_polars(bet_disks) -> NoReturn:
+    for index, disk in enumerate(bet_disks):
+        sectional_radiuses = disk.sectional_radiuses
+        sectional_polars = disk.sectional_polars
+        if len(sectional_radiuses) != len(sectional_polars):
+            raise ValueError(
+                f"In BET Disk {index}, length of sectional_radiuses ({len(sectional_radiuses)}) is not the same as that of sectional_polars ({len(sectional_polars)})."
```
[the remark this answers is not in the document]
Comment (feilin-flexcompute) [reply flow360/component/flow360_params/validations.py]: done

File: tests/params/test_validator_bet_disks.py (hunk 4 above)
Comment (feilin-flexcompute) [reply tests/params/test_validator_bet_disks.py]: left for a future PR

**Review by maciej-flexcompute — APPROVED** [review]

**Pull request merged**